File: app/src/main/java/net/lyxnx/carcheck/adapter/CardItemRecyclerAdapter.java
package net.lyxnx.carcheck.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TableLayout;
import android.widget.TextView;

import net.lyxnx.carcheck.R;
import net.lyxnx.carcheck.model.CardItem;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class CardItemRecyclerAdapter extends RecyclerView.Adapter<CardItemRecyclerAdapter.ViewHolder> {

    private final Context context;
    private List<CardItem> items;

    public CardItemRecyclerAdapter(Context context) {
        this.context = context;
    }

    @NonNull
    @Override
    public CardItemRecyclerAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(context).inflate(R.layout.info_card, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull CardItemRecyclerAdapter.ViewHolder holder, int position) {
        CardItem item = items.get(position);

        holder.header.setText(item.getHeader());

        item.getItems().entrySet().stream()
                .map(e -> {
                    View view = LayoutInflater.from(context).inflate(R.layout.row_layout, null, false);

                    ((TextView) view.findViewById(R.id.rowHeader)).setText(e.getKey());
                    ((TextView) view.findViewById(R.id.rowValue)).setText(e.getValue());

                    return view;
                })
                .forEach(view -> holder.table.addView(view));
    }

    @Override
    public int getItemCount() {
        return items == null ? 0 : items.size();
    }

    public void setItems(List<CardItem> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        TextView header;
        TableLayout table;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            header = itemView.findViewById(R.id.cardTitle);
            table = itemView.findViewById(R.id.table);
        }
    }
}

File: app/src/main/java/net/lyxnx/carcheck/adapter/CardItemRecyclerAdapter.kt
package net.lyxnx.carcheck.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import net.lyxnx.carcheck.R
import net.lyxnx.carcheck.widgets.InfoTableRow

class CardItemRecyclerAdapter : RecyclerView.Adapter<CardItemRecyclerAdapter.ViewHolder>() {

    var items: List<MoreInfoCardItem>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.info_card, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = items!![position]

        holder.header.text = item.category

        item.items.filterNotNull().forEach { (header, value) ->
            val row = InfoTableRow(holder.itemView.context)

            row.setHeader(header)
            row.setValue(value)

            holder.table.addView(row)
        }
    }

    override fun getItemCount() = items?.size ?: 0

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val header: TextView = itemView.findViewById(R.id.cardTitle)
        val table: TableLayout = itemView.findViewById(R.id.table)
    }
}

data class MoreInfoCardItem(val category: String, val items: List<Pair<String, String>?>)